src/radical/pilot/task_manager.py

```diff
@@ -695,19 +695,24 @@ def add_pilots(self, pilots):
         if len(pilots) == 0:
             raise ValueError('cannot add no pilots')
 
+        pilot_docs = list()
         with self._pilots_lock:
 
             # sanity check, and keep pilots around for inspection
             for pilot in pilots:
-                pid = pilot.uid
-                if pid in self._pilots:
-                    raise ValueError('pilot %s already added' % pid)
-                self._pilots[pid] = pilot
 
-                # subscribe for state updates
-                pilot.register_callback(self._pilot_state_cb)
+                if isinstance(pilot, dict):
+                    pilot_dict = pilot
+                else:
+                    pilot_dict = pilot.as_dict()
+                    # real object: subscribe for state updates
+                    pilot.register_callback(self._pilot_state_cb)
```

Comment:
If we pass a pilot object it also registers the callback, but not when we pass a dict. Does this mean that when an application passes a dict, it has to make sure to register TMGR's pilot callback?

Reply:
That is correct.  The callback is, after all, on the pilot object which we simply do not have in that case.  The rationale is that the API should normally use an object, and that is what we document.  Using the dict only happens to work when you know RP internals - but in that case you have other means to get state notifications (like, subscribing to the state pubsub).